### ISSUE TITLE
Why don't DefaultBandwidthMeter's reported elapsedMs and EventLogger's reported loadDurationMs match? 

### ISSUE THREAD

**Issue by emdash-ie** [google/ExoPlayer#3042]
Working in the demo application, I registered an event listener on the bandwidth meter, and then logged the provided elapsedMs values. I also logged the value of loadDurationMs provided to the event logger in onLoadCompleted.

Why don't these values match? The bandwidth listener's reported values seem to always be lower than those reported to the event logger, varying from about 50–100ms lower.

Here's my event listener:

```
public class BandwidthListener implements BandwidthMeter.EventListener {
    private static final String TAG = "BandwidthListener";

    @Override
    public void onBandwidthSample(int elapsedMs, long bytes, long bitrate) {
        Log.d(TAG, "elapsedMs: " + elapsedMs);
        Log.d(TAG, "bytes: " + bytes);
        Log.d(TAG, "Sampled bitrate (Mb) = " + (double) bytes * 8 / (double) elapsedMs / 1000);
        Log.d(TAG, "Estimated bitrate (Mb) = " + bitrate / 1E6);
    }
}
```

This is the code I have included in the EventLogger class:

```
  @Override
  public void onLoadCompleted(DataSpec dataSpec, int dataType, int trackType, Format trackFormat,
                              int trackSelectionReason, Object trackSelectionData, long mediaStartTimeMs,
                              long mediaEndTimeMs, long elapsedRealtimeMs, long loadDurationMs, long bytesLoaded) {
    Log.d(TAG, "Load duration: " + Long.toString(loadDurationMs));
  }
```

**Comment by ojw28** [comment]
What type of media are you playing?

**Comment by emdash-ie** [comment]
I'm playing the Google Play video in the YouTube DASH category in the demo app.

**Comment by ojw28** [comment]
For DASH streams audio and video are demuxed into separate streams, whose segments are downloaded in parallel. So if time is going from left to right in the diagram below, you might end up with loads overlapping as follows:
```
<---a1---><---a2---><---a3--->
<------v1-------><-----v2---->
```
The bandwidth estimate is updated whenever any load ends, so you'll get an updated estimate at the end of a1, then another at the end of v1 and so on. The period of time covered by the event at the end of v1 covers only the period of time between the end of a1 and the end of v1. Conversely, the `onLoadCompleted` corresponding to the end of v1 will report the period of time from the start of v1 to the end of v1.

So in short, you just shouldn't expect them to be the same.

**Comment by emdash-ie** [comment]
I see, thanks for the explanation – I appreciate it.

I also have a DASH stream that only contains video tracks, and for that stream the times still differ by 10–40 ms. What's happening there?

**Comment by ojw28** [comment]
I think that's a result of the thread that the timing is done on in the two cases. To start a load, the playback thread posts a message to a background thread that then does the load. When the load completes, the background thread posts a message back to the playback thread to say that it's finished. The timing for the bandwidth events is done directly on the background thread, where-as the timing for the loading events is done on the playback thread. So the playback thread timings are most likely slightly larger, since they also include the time taken to pass the two messages between the threads.

**Comment by emdash-ie** [comment]
I understand – thanks for your help.
